File: src/main/java/org/hoohoot/homelab/manager/notifications/kafka/RadarrNotificationsConsumer.java
package org.hoohoot.homelab.manager.notifications.kafka;

import io.smallrye.mutiny.Uni;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import jakarta.enterprise.context.ApplicationScoped;
import org.eclipse.microprofile.reactive.messaging.Incoming;
import org.eclipse.microprofile.rest.client.inject.RestClient;
import org.hoohoot.homelab.manager.notifications.matrix.MatrixAPI;
import org.hoohoot.homelab.manager.notifications.matrix.MatrixMessage;
import org.hoohoot.homelab.manager.notifications.matrix.MatrixRoomsConfiguration;

import java.util.UUID;

@ApplicationScoped
public class RadarrNotificationsConsumer {
    private final MatrixAPI matrixAPI;
    private final MatrixRoomsConfiguration matrixRooms;

    public RadarrNotificationsConsumer(@RestClient MatrixAPI matrixAPI, MatrixRoomsConfiguration matrixRooms) {
        this.matrixAPI = matrixAPI;
        this.matrixRooms = matrixRooms;
    }

    @Incoming("radarr-notifications")
    public Uni<Void> process(JsonObject payload) {
        JsonObject movie = payload.getJsonObject("movie");

        if (movie == null) return Uni.createFrom().voidItem();

        String title = movie.getString("title");
        String year = movie.getString("year");
        String imdbId = movie.getString("imdbId");


        JsonObject movieFile = payload.getJsonObject("movieFile");
        String quality = movieFile == null ? "" : movieFile.getString("quality");

        JsonArray tags = movie.getJsonArray("tags");

        String notificationContent = "<h1>Movie Downloaded</h1>" +
                                     "<p>" +
                                     "%s (%s) [%s] https://www.imdb.com/title/%s/<br>".formatted(title, year, quality, imdbId) +
                                     "Requested by : %s".formatted(requester(tags)) +
                                     "</p>";

        return this.matrixAPI.sendMessage(this.matrixRooms.radarr(), UUID.randomUUID().toString(), MatrixMessage.html(notificationContent)).replaceWithVoid();
    }

    @Deprecated
    public static String requester(JsonArray tags) {
        if (tags == null) return "unknown";

        return tags.stream()
                .map(Object::toString)
                .filter(tag -> tag.matches("\\d+ - \\w+"))
                .map(tag -> tag.split(" - ")[1])
                .findFirst()
                .orElse("unknown");
    }
}

File: src/main/kotlin/org/hoohoot/homelab/manager/notifications/kafka/RadarrNotificationsConsumer.kt
package org.hoohoot.homelab.manager.notifications.kafka

import io.smallrye.mutiny.Uni
import io.vertx.core.json.JsonArray
import io.vertx.core.json.JsonObject
import jakarta.enterprise.context.ApplicationScoped
import org.eclipse.microprofile.reactive.messaging.Incoming
import org.eclipse.microprofile.rest.client.inject.RestClient
import org.hoohoot.homelab.manager.notifications.matrix.MatrixAPI
import org.hoohoot.homelab.manager.notifications.matrix.MatrixMessage.Companion.html
import org.hoohoot.homelab.manager.notifications.matrix.MatrixRoomsConfiguration
import java.util.*

class RadarrNotificationsConsumer(
    @param:RestClient private val matrixAPI: MatrixAPI,
    private val matrixRooms: MatrixRoomsConfiguration
) {
    @Incoming("radarr-notifications")
    fun process(payload: JsonObject): Uni<Void> {
        val movie = payload.getJsonObject("movie") ?: return Uni.createFrom().voidItem()

        val title = movie.getString("title")
        val year = movie.getString("year")
        val imdbId = movie.getString("imdbId")


        val movieFile = payload.getJsonObject("movieFile")
        val quality = if (movieFile == null) "" else movieFile.getString("quality")

        val tags = movie.getJsonArray("tags")

        val notificationContent = "<h1>Movie Downloaded</h1>" +
                "<p>" +
                "%s (%s) [%s] https://www.imdb.com/title/%s/<br>".format(title, year, quality, imdbId) +
                "Requested by : %s".format(requester(tags)) +
                "</p>"

        return matrixAPI.sendMessage(
            matrixRooms.radarr(),
            UUID.randomUUID().toString(),
            html(notificationContent)
        ).replaceWithVoid()
    }

    companion object {
        @Deprecated("")
        fun requester(tags: JsonArray?): String {
            if (tags == null) return "unknown"

            return tags.stream()
                .map { obj: Any -> obj.toString() }
                .filter { tag: String -> tag.matches("\\d+ - \\w+".toRegex()) }
                .map { tag: String -> tag.split(" - ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[1] }
                .findFirst()
                .orElse("unknown")
        }
    }
}
